### PR TITLE
Fix browse category navigation

### PR DESCRIPTION
**Pull request by sheriffjimoh**
## Summary by Sourcery

Fix the browse category navigation by correctly handling active categories and chatbots. Improve URL path matching and navigation logic in various components for better accuracy and error handling. Enhance the ChatAccordion component to prevent unintended modal openings on profile pages.

Bug Fixes:
- Fix the browse category navigation by ensuring active categories and chatbots are correctly handled in the fetchThreads function.

Enhancements:
- Improve the URL path matching logic in SidebarLink and ChatbotComponent to use regular expressions for better accuracy.
- Refactor the handleAccordionToggle function in ThreadComponent to improve navigation logic and error handling.
- Enhance the ChatAccordion component to prevent opening the main thread modal on profile pages.

## Release Notes

- **New Features**
	- Enhanced URL building capabilities with structured error handling for various URL types.
	- Improved thread fetching logic in the `BrowseList` component based on user-selected categories and chatbots.
	- Updated navigation functionality in the `SidebarLink` and `ChatAccordion` components for better user experience.

- **Bug Fixes**
	- Resolved issues related to double-click behavior in the `SidebarLink` component.

- **Improvements**
	- Simplified pathname handling and thread management in the `Sidebar` component.
	- Streamlined component structure in the `ResponsiveSidebar` function.

- **Documentation**
	- Added type definitions for new functionalities in the `use-mb-chat` hook.

### CODE REVIEW

**Comment by AndlerRL** [comment]
[Fix Browse Sidebar & Page Navigation](https://trello.com/c/54RpZyon/272-fix-browse-sidebar-page-navigation)

**Review by max12-max — APPROVED** [review]

File: apps/masterbots.ai/components/layout/sidebar/sidebar-link.tsx [hunk 1]
```diff
@@ -139,8 +142,8 @@ export default function SidebarLink({ category, isFilterMode, page }: SidebarLin
 
   return (
     <div className={cn('flex flex-col mb-2')}>
-      <a
-       href="#"
+      <div
```
Comment (AndlerRL) [inline apps/masterbots.ai/components/layout/sidebar/sidebar-link.tsx:145]: Please re-consider this update... Remember `divitis`? Please avoid it. Use instead event listeners to block any activity inside, or simply add ARIA attributes to identify this `div` as an item of navigation. Additionally (and possibly the simplest way possible), you can also try to [passHref](https://nextjs.org/docs/pages/api-reference/components/link#passhref) when both links collide due to the nesting, i.e. if the `categoryContent` has a `anchor` (link) tags.

**Review by AndlerRL — CHANGES_REQUESTED** [review]
@sheriffjimoh -- I am concerned with the tests before submitting a PR and mark them as ready. I found regressions on navigation and the issue is not 100% fixed. I had to run some tests to make sure and you are not following the browse UI/UX in navigation. This is not the first time that I mention this to you. The profile thread navigation is not the same as browse in terms of UI/UX. If you have question what it is about, you can DM me or see the details on the video, so as previous conversations (I sent an audio with more details).

### Regression Issue: Chat Navigation Does Not Update the Thread List, only the active bot.

> ⚠️ Please, avoid `useCallback` and `useMemo`. Many times is just an array dependency on a useEffect... adding memoizing hooks hurts the performance and it is not needed on newer versions of react and nextjs. Adding also a function to a `useEffect` might also affect performance.

https://github.com/user-attachments/assets/94aef907-3d2b-4c88-9921-01a96aef734d

### Issue still remains, but with fewer chances.

> ⚠️ We are missing a `clean-up` return function call on a `useEffect` to reset the state whenever we go back on navigation.

https://github.com/user-attachments/assets/80ceba5f-4b29-49a3-bad6-146aa81d3c19

---

Besides the above, everything looks good to go, nice regExp strategy! Only make sure to test the sidebar navigation before pushing and all cases are covered.

**Comment by AndlerRL** [comment]
@sheriffjimoh -- look, this might be helpful for the back navigation, the [`router.events`](https://nextjs.org/docs/pages/api-reference/functions/use-router#routerevents) from NextJS.

image

File: apps/masterbots.ai/components/routes/chat/chat-accordion.tsx [hunk 2]
```diff
@@ -52,12 +59,22 @@ export const ChatAccordion = ({
...
+      if (profilePage) {
+      const category = thread?.chatbot?.categories[0]?.category?.name
+      const chatbot = thread?.chatbot?.name
+      const slug = params.slug;
+      if (!category || !chatbot || !slug) {
+            console.error('Missing required navigation parameters');
+            return;
+       }
+       router.push(`/u/${slug}/t/${toSlug(category)}/${toSlug(chatbot)}/${thread.threadId}`)
+      }
```
[the remark this answers is not in the document]
Comment (AndlerRL) [reply apps/masterbots.ai/components/routes/chat/chat-accordion.tsx]: @sheriffjimoh -- please follow this feedback.

**Comment by AndlerRL** [comment]
@sheriffjimoh -- let me know when it is ready for review :eyes: I still see the regression issue in the chat page.

**Comment by sheriffjimoh** [comment]
> @sheriffjimoh -- let me know when it is ready for review 👀 I still see the regression issue in the chat page.

Hmm, yeah i only did it for profile page,  just seen that it's happening on the chat page as well, the  use of router you suggested does  not seems to work since i can't use `next/router` on next again and `next/navigation` does not have those events method. also, we can not depend on the route or pathname change only too,  it will affect  the popup when it's really needed, especially when you click it to show on the browse page you know that update the pathname too. either way i will fix that.

File: apps/masterbots.ai/lib/url.ts [hunk 3]
```diff
@@ -16,4 +17,83 @@ export const encodeQuery = (input: string): string => {
...
+        const missing = Object.entries(params)
+          .filter(([_, value]) => !value)
+          .map(([key]) => key)
+          .join(', ');
+          
+        console.error(`Missing required parameters for thread URL: ${missing}`);
+        return '/';
+      }
+
+      return `/u/${encodeURIComponent(slug)}/t/${toSlug(category)}/${toSlug(chatbot)}/${threadId}`;
```
[the remark this answers is not in the document]
Comment (AndlerRL) [reply apps/masterbots.ai/lib/url.ts:55]: @coderabbitai make an issue to track this and assign @sheriffjimoh to it. Additionally, include a difficulty tag on the issue.

File: apps/masterbots.ai/lib/url.ts [hunk 4]
```diff
@@ -16,4 +17,83 @@ export const encodeQuery = (input: string): string => {
...
+    } catch (error) {
+      console.error('Error constructing thread URL:', error);
+      return '/';
+    }
+  },
```
[the remark this answers is not in the document]
Comment (AndlerRL) [reply apps/masterbots.ai/lib/url.ts:60]: @coderabbitai it is better to deal with this with another PR. Let's track this into an issue but in general context:

We need to create a set of error code with their types like the given class as example. Analyze which parts needs error handling and provide steps to produce this improvement.

**Review by AndlerRL — APPROVED** [review]
LGTM! I experienced smooth navigation between categories (domains) in browse and chat categories (domains) and chatbots. I added some comments for coderabbitai to track new issues regarding to the given feedback. Some of them I resolved them. I will be resolving those conversations once the Ai respond.

Merging!

**Pull request merged**